File: resources/expected/ros2/fan_in_fan_out/strict/src/fan_in_fan_out_system_cpp_pkg/src/base_code/fanIn_consumer_base_src.cpp
#include "fan_in_fan_out_system_cpp_pkg/base_headers/fanIn_consumer_base_src.hpp"

//========================================================
// Re-running Codegen will overwrite changes to this file
//========================================================

fanIn_consumer_base::fanIn_consumer_base() : Node("fanIn_consumer")
{
    cb_group_ = this->create_callback_group(rclcpp::CallbackGroupType::Reentrant);
    subscription_options_.callback_group = cb_group_;

    // Setting up connections
    fanIn_consumer_myInteger_subscription_ = this->create_subscription<fan_in_fan_out_system_cpp_pkg_interfaces::msg::Integer64>(
        "fanIn_consumer_myInteger",
        1,
        std::bind(&fanIn_consumer_base::accept_myInteger, this, std::placeholders::_1), subscription_options_);

    // Used by receiveInputs
    inDataPortTupleVector = {
    };

    // Used by sendOutputs
    outPortTupleVector = {
    };

}

//=================================================
//  C o m m u n i c a t i o n
//=================================================

void fanIn_consumer_base::accept_myInteger(fan_in_fan_out_system_cpp_pkg_interfaces::msg::Integer64 msg)
{
    enqueue(infrastructureIn_myInteger, msg);
    std::thread([this]() {
        std::lock_guard<std::mutex> lock(mutex_);
        receiveInputs(infrastructureIn_myInteger, applicationIn_myInteger);
        if (applicationIn_myInteger.empty()) return;
        handle_myInteger_base(applicationIn_myInteger.front());
        applicationIn_myInteger.pop();
        sendOutputs();
    }).detach();
}

void fanIn_consumer_base::handle_myInteger_base(MsgType msg)
{
    if (auto typedMsg = std::get_if<fan_in_fan_out_system_cpp_pkg_interfaces::msg::Integer64>(&msg)) {
        handle_myInteger(*typedMsg);
    } else {
        PRINT_ERROR("Receiving wrong type of variable on port myInteger.\nThis shouldn't be possible.  If you are seeing this message, please notify this tool's current maintainer.");
    }
}

void fanIn_consumer_base::receiveInputs(std::queue<MsgType>& infrastructureQueue, std::queue<MsgType>& applicationQueue) {
    if (!infrastructureQueue.empty()) {
        MsgType eventMsg = infrastructureQueue.front();
        infrastructureQueue.pop();
        enqueue(applicationQueue, eventMsg);
    }

    for (std::tuple<std::queue<MsgType>*, std::queue<MsgType>*> port : inDataPortTupleVector) {
        auto infrastructureQueue = std::get<0>(port);
        if (!infrastructureQueue->empty()) {
            auto msg = infrastructureQueue->front();
            enqueue(*std::get<1>(port), msg);
        }
    }
}

void fanIn_consumer_base::enqueue(std::queue<MsgType>& queue, MsgType val) {
    if (queue.size() >= 1) {
        queue.pop();
    }
    queue.push(val);
}

void fanIn_consumer_base::sendOutputs() {
    for (std::tuple<std::queue<MsgType>*, std::queue<MsgType>*, void (fanIn_consumer_base::*)(MsgType)> port : outPortTupleVector) {
        auto applicationQueue = std::get<0>(port);
        if (applicationQueue->size() != 0) {
            auto msg = applicationQueue->front();
            applicationQueue->pop();
            enqueue(*std::get<1>(port), msg);
        }
    }

    for (std::tuple<std::queue<MsgType>*, std::queue<MsgType>*, void (fanIn_consumer_base::*)(MsgType)> port : outPortTupleVector) {
        auto infrastructureQueue = std::get<1>(port);
        if (infrastructureQueue->size() != 0) {
            auto msg = infrastructureQueue->front();
            infrastructureQueue->pop();
            (this->*std::get<2>(port))(msg);
        }
    }
}

File: resources/expected/ros2/python-fan_in_fan_out_inverted_topics/strict/src/fan_in_fan_out_system_py_pkg/fan_in_fan_out_system_py_pkg/base_code/fanIn_consumer_base_src.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from collections import deque
from typing import Union
import threading
from rclpy.callback_groups import ReentrantCallbackGroup
from fan_in_fan_out_system_py_pkg_interfaces.msg import Integer64

#========================================================
# Re-running Codegen will overwrite changes to this file
#========================================================

class fanIn_consumer_base(Node):
    def __init__(self):
        super().__init__("fanIn_consumer")

        self.cb_group_ = ReentrantCallbackGroup()

        self.lock_ = threading.Lock()

        # Setting up connections
        self.fanIn_consumer_myInteger_subscription_1 = self.create_subscription(
            Integer64,
            "fanIn_producer1_myInteger",
            self.accept_myInteger,
            1,
            callback_group=self.cb_group_)

        self.fanIn_consumer_myInteger_subscription_2 = self.create_subscription(
            Integer64,
            "fanIn_producer2_myInteger",
            self.accept_myInteger,
            1,
            callback_group=self.cb_group_)

        self.infrastructureIn_myInteger = deque()
        self.applicationIn_myInteger = deque()

        # Used by receiveInputs
        self.inDataPortTupleVector = [
        ]

        # Used by sendOutputs
        self.outPortTupleVector = [
        ]

    def timeTriggered(self):
        raise NotImplementedError("Subclasses must implement this method")

    #=================================================
    #  C o m m u n i c a t i o n
    #=================================================

    def myInteger_thread(self):
        with self.lock_:
            self.receiveInputs(self.infrastructureIn_myInteger, self.applicationIn_myInteger)
            if len(self.applicationIn_myInteger) == 0: return
            self.handle_myInteger_base(self.applicationIn_myInteger[0])
            self.applicationIn_myInteger.pop()
            self.sendOutputs()

    def accept_myInteger(self, msg):
        self.enqueue(self.infrastructureIn_myInteger, msg)
        thread = threading.Thread(target=self.myInteger_thread)
        thread.daemon = True
        thread.start()


    #=================================================
    #  C o m p u t e    E n t r y    P o i n t
    #=================================================
    def handle_myInteger(self, msg):
        raise NotImplementedError("Subclasses must implement this method")

    def handle_myInteger_base(self, msg):
        if type(msg) is Integer64:
            self.handle_myInteger(msg)
        else:
            self.get_logger.error("Receiving wrong type of variable on port myInteger.\nThis shouldn't be possible.  If you are seeing this message, please notify this tool's current maintainer.")


    def receiveInputs(self, infrastructureQueue, applicationQueue):
        if not(len(infrastructureQueue) == 0):
            eventMsg = infrastructureQueue[0]
            infrastructureQueue.pop()
            self.enqueue(applicationQueue, eventMsg)

        for port in self.inDataPortTupleVector:
            infrastructureQueue = port[0]
            if not(len(infrastructureQueue) == 0):
                msg = infrastructureQueue[0]
                self.enqueue(port[1], msg)

    def enqueue(self, queue, val):
        if len(queue) >= 1:
            queue.pop()
        queue.append(val)

    def sendOutputs(self):
        for port in self.outPortTupleVector:
            applicationQueue = port[0]
            if len(applicationQueue) != 0:
                msg = applicationQueue[0]
                applicationQueue.pop()
                self.enqueue(port[1], msg)

        for port in self.outPortTupleVector:
            infrastructureQueue = port[1]
            if len(infrastructureQueue) != 0:
                msg = infrastructureQueue[0]
                infrastructureQueue.pop()
                (port[2])(msg)
